### DISCUSSION TITLE
Field 'value' might be missing on 'currentValue' message

### DISCUSSION THREAD

**Discussion by s-frei** [zwave-js/zwave-js-ui#2506] [General]
Hi there,

first of all I really appreaciate your good work you are doing here, and to
see the progress over the last years! 🚀  Keep up the good work.

---

Now to the problem I encountered just recently. I'm deserializing the *Entire Z-Wave value Object* messages
to corresponding Java objects. I've unit tested those deserializations with *real* collected messages using `v5.8.0`.
The `type` field in the messages are giving me a hint what type of object I could expect, parsing the `value`.

This works well for a message like (used in unit test):
```json
{
  "id": "3-38-1-currentValue",
  "nodeId": 3,
  "commandClass": 38,
  "commandClassName": "Multilevel Switch",
  "endpoint": 1,
  "property": "currentValue",
  "propertyName": "currentValue",
  "type": "number",
  "readable": true,
  "writeable": false,
  "label": "Current value",
  "stateless": false,
  "min": 0,
  "max": 99,
  "list": false,
  "value": 99,
  "isCurrentValue": true,
  "targetValue": "38-1-targetValue",
  "lastUpdate": 1627285239197
}
```
Today I got some stacktraces on deserializing messages sent with `v6.12.0`. Example of failing message:
```json
{
  "id": "3-38-2-currentValue",
  "nodeId": 3,
  "commandClass": 38,
  "commandClassName": "Multilevel Switch",
  "endpoint": 2,
  "property": "currentValue",
  "propertyName": "currentValue",
  "type": "number",
  "readable": true,
  "writeable": false,
  "label": "Current value",
  "stateless": false,
  "commandClassVersion": 4,
  "min": 0,
  "max": 99,
  "list": false,
  "isCurrentValue": true,
  "targetValue": "38-2-targetValue",
  "lastUpdate": 1656661337621
}
```
because the `value` field is missing there. Is it expectable that this field might be missing from now on?
Is this related to the change made here https://github.com/zwave-js/zwavejs2mqtt/commit/d437d5e359dda7a2a4f895f10ec8cac3917c78bd?

The actual used device:
| Manufacturer | Product            | Product code |
|:------------:|:------------------:|:------------:|
| Fibargroup   |  Roller Shutter 3  | FGR223       |

Couldn't find anything more specific about this behaviour and wasn't sure if this should be an *issue*.

If you need more information, additional data, and so on, let me know. I'll try to response as fast as
possible 🙂

**Comment by robertsLando** [comment]
@s-frei version 5.8 is from 8 months ago, lot of things have changed in the meanwhile

> because the value field is missing there. Is it expectable that this field might be missing from now on?

The value shouldn't be missing, at least I didn't change that part of code that handles it: https://github.com/zwave-js/zwavejs2mqtt/blob/master/lib/Gateway.ts#L1805

Does the value is ALWAYS missing or just sometime? Did you tried to change the payload type to see if value is present with JSON/Raw type?

**Reply by s-frei** [reply]
Hi @robertsLando,

thank you for your fast response!

> @s-frei version 5.8 is from 8 months ago, lot of things have changed in the meanwhile

Yes, I noticed that and I'm currently preparing an upgrade.

> Does the value is ALWAYS missing or just sometime? Did you tried to change the payload type to see if value is present with JSON/Raw type?

What exactly to you mean with *raw*? I tried all applicable payload types:
- JSON Time-Value
- Entire Z-Wave value Object
- Just value (not applicable because using Hass discovery)

the value is always missing. 

When hitting *DEBUG INFO* for the node I also get:

```json
{
  "id": "3-38-2-currentValue",
  "nodeId": 3,
  "commandClass": 38,
  "commandClassName": "Multilevel Switch",
  "endpoint": 2,
  "property": "currentValue",
  "propertyName": "currentValue",
  "type": "number",
  "readable": true,
  "writeable": false,
  "label": "Current value",
  "stateless": false,
  "commandClassVersion": 4,
  "min": 0,
  "max": 99,
  "list": false,
  "isCurrentValue": true,
  "targetValue": "38-2-targetValue",
  "lastUpdate": 1656682219286
}
```

For debugging I currently use [MQTTExplorer](https://github.com/thomasnordquist/MQTT-Explorer) which has a *history* feature for messages and in each message for the topic `3/38/2/currentValue` the `value` is missing.

Maybe it is relevant that the UI shows:
<img width="1218" alt="grafik" src="https://user-images.githubusercontent.com/25681388/176907161-ed5a1de8-3580-4257-9cb6-28788ea77d17.png">

Please also let me know if I can help you better, somehow.
I could also dive into some code if you point me a direction 🙂

**Reply by robertsLando** [reply]
It could be that the current value is not defined for some reason, I have many devices with Multilevel Switch CC and they all have currentValue set, try to press on refresh values button you find in top right corner of each expandable command class container

**Reply by robertsLando** [reply]
Also it could be worth doing a re-interview of all your nodes

**Reply by s-frei** [reply]
Well, I forgot to maybe tell you, that the first instance is resolved correctly:
<img width="1221" alt="grafik" src="https://user-images.githubusercontent.com/25681388/176918201-3945c2d9-ff4c-4c8a-989a-39cdabf92fdd.png">
so I will hit the *Refresh* now.

**Reply by s-frei** [reply]
Nonethless I'll handle the possibility of the missing `value` field.

**Reply by s-frei** [reply]
Ok, refreshing the Multilevel Switch CC did the job:
```text
zwavejs2mqtt    | 14:54:20.783 CNTRLR �� [Node 003] requesting current switch state...                     [Endpoint 1]
zwavejs2mqtt    | 2022-07-01 14:54:21.250 INFO ZWAVE: Node 3: value updated: 38-1-targetValue 90 => 90
zwavejs2mqtt    | 2022-07-01 14:54:21.255 INFO ZWAVE: Node 3: value updated: 38-1-duration unknown => unknown
zwavejs2mqtt    | 2022-07-01 14:54:21.258 INFO ZWAVE: Node 3: value updated: 38-1-currentValue 90 => 90
zwavejs2mqtt    | 14:54:21.269 CNTRLR �� [Node 003] requesting current switch state...                     [Endpoint 2]
zwavejs2mqtt    | 2022-07-01 14:54:21.703 INFO ZWAVE: Node 3: value updated: 38-2-targetValue 10 => 99
zwavejs2mqtt    | 2022-07-01 14:54:21.710 INFO ZWAVE: Node 3: value added: 38-2-duration => unknown
zwavejs2mqtt    | 2022-07-01 14:54:21.711 INFO ZWAVE: Node 3: value added 3-38-2-duration => unknown
zwavejs2mqtt    | 2022-07-01 14:54:21.714 INFO ZWAVE: Node 3: value added: 38-2-currentValue => 99
zwavejs2mqtt    | 2022-07-01 14:54:21.773 INFO ZWAVE: Node 3: value added 3-38-2-currentValue => 99
zwavejs2mqtt    | 2022-07-01 14:54:21.782 INFO ZWAVE: Success zwave api call refreshCCValues undefined
```
<img width="1237" alt="grafik" src="https://user-images.githubusercontent.com/25681388/176919243-bb12ba0e-ac9c-471a-8f56-5f268afdaad1.png">

thank you for your help.

Maybe it would be an improvement always having a `value` field (nullish), even though it hasn't got a value.

**Reply by robertsLando** [reply]
> Maybe it would be an improvement always having a value field (nullish), even though it hasn't got a value.

Values should never be undefined, if so it means there is an error that should be fixed (like with a refresh). This could happen maybe when driver is updated and there is some kind of inconsistency in cache. Nothing so bas BTW

**Reply by s-frei** [reply]
My intention here was to keep the API (JSON via MQTT) consistent and stick to the displayed value on the UI.
It's always nice to know exactly what to expect. We've got multiple gateways, so maybe I need to implement the *refreshInfo(nodeId)* API.

But anyhow, thank you for your help on this one! If you want to, let me know if I/we could help you somewhere and I'll try to get some time for this. Maybe something UI or doumentation related 🙂

**Reply by robertsLando** [reply]
the problem is that when a value is `undefined` in a JS object and you convert it to JSON string to send as MQTT payload it will strip out undefined props, I could set it to null to prevent it if needed

**Reply by s-frei** [reply]
I see... For me it's okay that the `value` might be missing and I'm able to handle this correctly now. I don't know the impact of adding the field with a *null* value when it can't be determined, for others. I personally think this would be a good idea since it makes clear what the currentValue is, even though it's *'nothing'*. This could als be an optional feature on the *Gateway* settings.
